osd/PrimaryLogPG: do not set data digest for bluestore

xiexingguo opened the pull request:
BlueStore enables CRC by default, so this is a dup and gains no more benefits.

Also default to no data/omap digest for new object.

branch-predictor commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -5750,12 +5750,17 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
-	  obs.oi.set_data_digest(osd_op.indata.crc32c(obs.oi.data_digest));
-	else
+	} else if (op.extent.offset == oi.size && obs.oi.is_data_digest()) {
+          if (g_conf->get_val<bool>("osd_skip_data_digest"))
```
You could move that out of "for" loop and use local boolean variable, as get_val() is quite expensive.

branch-predictor commented:
What about filestore? It shouldn't be default there.

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -4937,6 +4937,8 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
   ObjectState& obs = ctx->new_obs;
   object_info_t& oi = obs.oi;
   const hobject_t& soid = oi.soid;
+  bool skip_data_digest = osd->store->get_type() == "bluestore" &&
```
probably this should be a property of the underlying store. i.e. we can just hard wire it to true for bluestore and false for filestore.

liu-chunmei commented:
looks good, please check why make check failed.

xiexingguo commented:
> looks good, please check why make check failed.

It's a known issue.

retest this please

tchaikov commented on src/common/options.cc:
```diff
@@ -1919,6 +1919,10 @@ std::vector<Option> get_global_options() {
     .set_default(8)
     .set_description(""),
 
+    Option("osd_skip_data_digest", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
```
i don't think this is necessary with the `has_internal_csum()` method. @liewegas and @jdurgin , what do you think?

xiexingguo replied:
Fine with me:-)

liewegas replied:
Yeah i don't mind either way honestly!

tchaikov commented on src/os/ObjectStore.h:
```diff
@@ -2030,6 +2030,9 @@ class ObjectStore {
   virtual void inject_mdata_error(const ghobject_t &oid) {}
 
   virtual void compact() {}
+  virtual bool has_internal_csum() {
```
could you s/internal/builtin/ and mark this method `const`

tchaikov replied:
> and mark this method `const`?

@xiexingguo could you address this also?

xiexingguo replied:
Sure. It slips over my eyes by accident...

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -5750,12 +5751,17 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
-	else
+	} else if (op.extent.offset == oi.size && obs.oi.is_data_digest()) {
+          if (skip_data_digest)
+            obs.oi.clear_data_digest();
```
if the object never sets the digest, why bother clearing it?

xiexingguo replied:
I intent to take the chance to cleanup any previous existing BlueStore data_digest since this change is going to land on mimic(post-luminous)?

tchaikov replied:
i understand you do this with good intention, but what is the merit of doing so? or put in other words, what the downside if we have an "oi.is_data_digest()" on a store with builtin checksum support? will we run into some trouble under such circumstance?

xiexingguo replied:
The problem is: if we keep **appending** an existing object  previously with data_digest on(instead of writing new objects or random writing), we'll be busy in generating new data_digests; hence the skip_data_digest has no effect and no benefits will be gained by applying this change

xiexingguo replied:
> will we run into some trouble under such circumstance?

Not really

tchaikov replied:
> if we keep appending an existing object previously with data_digest on(instead of writing new objects or random writing), we'll be busy in generating new data_digests; 

ahh, right! i thought we skipped this if the store has builtin checksum support.

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -5750,12 +5751,17 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
-	  obs.oi.set_data_digest(osd_op.indata.crc32c(obs.oi.data_digest));
-	else
+	} else if (op.extent.offset == oi.size && obs.oi.is_data_digest()) {
+          if (skip_data_digest)
```
i see you are adding `{` and `}` for if blocks, but why do you make this one an exception?

liewegas commented:
1. We should be aware that csum on bluestore can be disabled, so it's possible that someone could turn of bluestore csum and want this on.  (I don't think this is worth worrying about, though.)
2. We need a similar change in the deep scrub code so that we don't calculate and set the digest, right?

yuriw commented:
http://pulpito.ceph.com/yuriw-2017-09-07_18:45:31-rados-wip-yuri-testing3-2017-09-07-1719-distro-basic-smithi/

the scrub.sh test needs fixing see http://pulpito.ceph.com/yuriw-2017-09-07_18:45:31-rados-wip-yuri-testing3-2017-09-07-1719-distro-basic-smithi/1607250/

xiexingguo commented:
> We should be aware that csum on bluestore can be disabled, so it's possible that someone could turn of bluestore csum and want this on. (I don't think this is worth worrying about, though.)

Introduce an extra option "osd_skip_data_digest" for that purpose.


> We need a similar change in the deep scrub code so that we don't calculate and set the digest, right?

Yeah, updated. Fixed the scrub test also (http://pulpito.ceph.com/xxg-2017-09-12_04:52:04-rados-wip-data-digest-distro-basic-smithi/).

xiexingguo commented:
retest this please

xiexingguo commented:
Another scrub-test fix(http://pulpito.ceph.com/yuriw-2017-09-12_21:09:21-rados-wip-yuri-testing2-2017-09-12-1629-distro-basic-smithi/)

Rerun at http://pulpito.ceph.com/xxg-2017-09-13_02:55:35-rados-wip-data-digest-distro-basic-smithi/ and passed

xiexingguo commented:
http://pulpito.ceph.com/yuriw-2017-09-12_21:09:21-rados-wip-yuri-testing2-2017-09-12-1629-distro-basic-smithi/
The failure one was addressed and reruned at
http://pulpito.ceph.com/xxg-2017-09-13_09:08:26-rados-wip-xxg-testing-2017-09-13-1322-distro-basic-smithi/
